bug(provider-runtime): bundled provider-runtime.mjs hangs at top-level startup, app stuck on "Reconnecting" forever

## What is observed (verified evidence)

The local provider runtime that powers Claude Code, Codex, and Gemini agents inside Seren Desktop fails to become ready. The Tauri side spawns the bundled node script `target/debug/embedded-runtime/provider-runtime/provider-runtime.mjs`, then polls `http://127.0.0.1:<ephemeral>/__seren/health` for 10 seconds in [src-tauri/src/provider_runtime.rs:293-331](src-tauri/src/provider_runtime.rs#L293-L331). Every poll attempt fails with connection refused, and the Rust side eventually returns the literal string `Timed out waiting for provider runtime readiness.`

Symptoms in the running app:
- Chat UI is stuck on "Reconnecting" forever
- DevTools console floods with `[Debug] reqwest::connect ... starting new connection: http://127.0.0.1:<port>/` every ~150ms
- Eventually `[Error] Failed to load available agents: "Timed out waiting for provider runtime readiness."` from `agent.store.ts:415`
- New port chosen on each retry (64820, 64889, 64978, 65053, 65193, 65270, ...) — that's the Rust side respawning the child each time the previous timeout fired
- `[Info] [ProviderRuntime] Found provider-runtime.mjs at "/Users/taariqlewis/Projects/Seren_Projects/seren-desktop/src-tauri/target/debug/embedded-runtime/provider-runtime/provider-runtime.mjs"` — so the Rust side IS finding the bundle, the child process IS being spawned, it just never produces a healthy `/__seren/health` response

## Direct test of the bundled script (the smoking gun)

Running the bundled `provider-runtime.mjs` directly from the terminal with the same embedded node binary the Rust side uses, on a fixed port:

```bash
/Users/taariqlewis/Projects/Seren_Projects/seren-desktop/src-tauri/target/debug/embedded-runtime/darwin-x64/node/bin/node \
  /Users/taariqlewis/Projects/Seren_Projects/seren-desktop/src-tauri/target/debug/embedded-runtime/provider-runtime/provider-runtime.mjs \
  --host 127.0.0.1 --port 56789 --token testtoken &
sleep 4
curl -sS http://127.0.0.1:56789/__seren/health
```

**Result**: zero stdout/stderr output for the entire 4 seconds, the process is still alive but unresponsive, `curl` returns `Failed to connect to 127.0.0.1 port 56789 after 0 ms: Couldn't connect to server`.

A second test with a 5-second timeout via a wrapper produced **zero bytes** of output. The script is hung at top-level execution, before reaching any HTTP server bind, before printing anything. It does not crash, it does not exit, it does not produce diagnostics — it just sits there.

## Scope of the hang

The hang is in JavaScript top-level execution of the bundle. Confirmed it's not the Rust side: [src-tauri/src/provider_runtime.rs:293-331](src-tauri/src/provider_runtime.rs#L293-L331) is async via `reqwest::Client` and is provably running its retry loop (the DevTools console shows the polling).

The bundled script `target/debug/embedded-runtime/provider-runtime/provider-runtime.mjs` imports:
```javascript
import { randomBytes } from "node:crypto";
import http from "node:http";
import { WebSocketServer } from "ws";
import { createProviderHandlers } from "./browser-local/providers.mjs";
import { addClient, emit, removeClient } from "./browser-local/events.mjs";
import { handleRpcMessage, registerHandler } from "./browser-local/rpc.mjs";

// (#1502 host marker — added today)
process.env.SEREN_HOST = "seren-desktop";
process.env.SEREN_DESKTOP = "1";
```

The hang is happening before line 17 (the `usage()` function definition) executes, because the script's first observable output is supposed to come from `parseArgs` / `startServer` calls that never run.

The local `browser-local/` directory contains:
- `agent-registry.mjs` (17KB)
- `claude-runtime.mjs` (50KB)
- `dialogs.mjs` (3.6KB)
- `events.mjs` (847B)
- `fs.mjs` (2.1KB)
- `gemini-runtime.mjs` (32KB)
- `mcp-config.mjs` (4KB)
- `providers.mjs` (39KB)
- `rpc.mjs` (1.5KB)

## What's been ruled out

- **Not #1501 / #1503 (`mcp::mcp_connect` fix).** That bug was specifically about `mcp::mcp_connect` doing a blocking `read_line` on a child's stdio on the main Tauri thread. PR #1503 fixed it via `tokio::task::spawn_blocking + tokio::time::timeout`. The fix is verified working: the same DevTools log shows `[ERROR] [MCP:playwright] MCP initialize handshake timed out after 15s — check that the server command is correct and the server emits a valid JSON-RPC response on stdout` — that exact string is the new error message from #1503, proving the fix is in place. The provider runtime hang is in completely different code paths.
- **Not #1498 / #1499 / #1500 (`#1492` claude memory interceptor).** None of those PRs touched `provider_runtime.rs`, `bin/provider-runtime.mjs`, or any of the `browser-local/*.mjs` modules. They only touched `claude_memory.rs`, `commands/claude_memory.rs`, `commands/memory.rs` (one read-only `base_url()` accessor), `lib.rs` (command registration), `App.tsx`, `SettingsPanel.tsx`, `claudeMemory.ts`, `settings.store.ts`, and one new vitest file.
- **Not the Rust polling code.** Verified by the log lines showing `[ProviderRuntime] Found provider-runtime.mjs at ...` followed by repeated `reqwest::connect` to different ports — that's the Rust loop running fine, the child just never serves the health endpoint.

## Suspect: PR #1502 (host marker)

PR #1502 (`feat(skills): add host marker and exclude-hosts contract for CLI-only skills`, commit `77bdf2b0`) is the only PR that landed today that touched files in this code path. It modified `bin/provider-runtime.mjs` (added two `process.env.SEREN_HOST = ...` lines at the top after the imports) and `src-tauri/src/lib.rs` (added two `unsafe { std::env::set_var(...) }` calls in the Tauri `setup` hook).

The two new lines in `bin/provider-runtime.mjs` look trivial and synchronous, but the PR must have triggered a rebuild of the entire `bin/browser-local/` bundle when `pnpm build:provider-runtime` ran. Since the script-level hang is BEFORE any code we wrote runs, the most likely cause is one of the imported `browser-local/*.mjs` modules has a top-level side effect that hangs (network call at module load, infinite loop, deadlocked promise, etc.) — possibly something that was inadvertently changed or rebundled differently.

The `unsafe { std::env::set_var }` in `lib.rs` is also worth a look — `set_var` is unsafe in Rust 2024 because it is not thread-safe, and the `setup` hook runs concurrently with other initialization. The PR's safety comment claims "called exactly once during app setup before any threads spawn child processes" but the Tauri setup hook runs after some plugin initialization, and there are multiple background tasks running by then. Could be racy.

## Reproduction (manual)

1. Check out current `main` (`dee8e8fe` or later)
2. `pnpm tauri dev`
3. Wait for app to launch and complete login
4. Observe: chat UI stuck on "Reconnecting"
5. DevTools console will show repeated `reqwest::connect` to `127.0.0.1:<changing port>` and `[Error] Failed to load available agents: "Timed out waiting for provider runtime readiness."`

Or, even simpler reproduction (no Tauri needed):

```bash
/Users/taariqlewis/Projects/Seren_Projects/seren-desktop/src-tauri/target/debug/embedded-runtime/darwin-x64/node/bin/node \
  /Users/taariqlewis/Projects/Seren_Projects/seren-desktop/src-tauri/target/debug/embedded-runtime/provider-runtime/provider-runtime.mjs \
  --host 127.0.0.1 --port 56789 --token testtoken
```

Expected: usage banner or HTTP server bind log within 1 second. Actual: zero output, process hung indefinitely.

## Investigation path the next engineer should follow

1. **Identify which import hangs.** Run a node script that imports each `browser-local/*.mjs` module individually with `console.log` between each, find the first one that doesn't return.
2. **Inspect that module's top-level code.** Look for accidental top-level `await`, network calls, unawaited promises, etc.
3. **Check the bundle build script.** `pnpm build:provider-runtime` is `tsx scripts/build-provider-runtime.ts`. Verify the bundling step itself didn't introduce a corruption — diff the bundled output against the source `bin/browser-local/*.mjs` files.
4. **Audit #1502's `unsafe { set_var }` in `lib.rs`.** Determine whether it's actually being called before any threads spawn child processes, or if there's a race with the embedded runtime spawning subprocesses concurrently.
5. **Verify the embedded node binary itself is healthy.** Run `<embedded-node> --version` and `<embedded-node> -e 'console.log("ok")'` to rule out a corrupt binary.

## Impact

**HIGH severity**: every Seren Desktop dev build is currently unable to run any agent (Claude Code, Codex, Gemini) because the provider runtime never becomes ready. The chat UI is unusable. This blocks:
- Any work that depends on the local agents
- The end-to-end smoke test for #1492 (claude memory interceptor) — the feature itself is shipped and unit-tested but cannot be verified through the dev UI until this is fixed
- Any new dev environment a contributor sets up

The installed `/Applications/SerenDesktop.app` (built from a previous main) does not have this issue, so this is dev-build only — but it's actively breaking the developer loop.

## Acceptance criteria
- [ ] `pnpm tauri dev` launches and the chat UI shows a normal connected state, not "Reconnecting"
- [ ] The bundled `provider-runtime.mjs` produces output and binds its HTTP server within 2 seconds when run directly with `node`
- [ ] The `/__seren/health` endpoint returns `{"ok": true}` on the assigned port
- [ ] The root cause is documented in the fix PR (which import / which code change introduced the hang)
- [ ] If the fix involves reverting any part of #1502, the host-marker contract from #1496 is preserved by an alternative mechanism

## Related
- Discovered during #1492 smoke test (claude memory interceptor)
- #1501 / #1503 are the MCP main-thread hang fix — a different bug with a similar shape but completely separate code path. Both are in place and working.
- #1502 (commit `77bdf2b0`) is the most recently merged PR touching `bin/provider-runtime.mjs` and `src-tauri/src/lib.rs`. Worth bisecting against.

--- a/scripts/prepare-mcp-servers.ts
+++ b/scripts/prepare-mcp-servers.ts
@@ -9,6 +9,8 @@ const ROOT = path.resolve(import.meta.dirname, "..");
 const SOURCE = path.join(ROOT, "mcp-servers", "playwright-stealth");
 const DEST_PARENT = path.join(ROOT, "src-tauri", "mcp-servers");
 const DEST = path.join(DEST_PARENT, "playwright-stealth");
+const DEST_NEW = path.join(DEST_PARENT, "playwright-stealth.new");
+const DEST_OLD = path.join(DEST_PARENT, "playwright-stealth.old");
 const RM_RETRY_OPTS = { recursive: true, force: true, maxRetries: 10, retryDelay: 100 } as const;
 
 // Main
@@ -30,14 +32,37 @@ execSync("pnpm install --node-linker=hoisted", { cwd: SOURCE, stdio: "inherit" }
 console.log("Building...");
 execSync("pnpm build", { cwd: SOURCE, stdio: "inherit" });
 
-// 3. Clean destination
-if (fs.existsSync(DEST_PARENT)) {
-	console.log("Cleaning destination...");
-	fs.rmSync(DEST_PARENT, RM_RETRY_OPTS);
+// 3. Atomic swap: copy to sibling, then rename into place.
+//
+// The previous implementation wiped DEST_PARENT then cpSync'd the source back,
+// leaving a ~1-3s window where src-tauri/mcp-servers/playwright-stealth/ did not
+// exist. Tauri's build script walks that tree on fresh builds to emit
+// `rerun-if-changed` entries for bundle resources and aborts with
+// "resource path doesn't exist" if a file vanishes mid-walk. Running both
+// concurrently via `beforeDevCommand` reliably raced on cold builds (#1505).
+//
+// Two renameSync calls on the same filesystem shrink that window to a few
+// microseconds, narrow enough that the walker will not observe a missing dest.
+fs.mkdirSync(DEST_PARENT, { recursive: true });
+
+if (fs.existsSync(DEST_NEW)) {
+	fs.rmSync(DEST_NEW, RM_RETRY_OPTS);
+}
+console.log("Staging new bundle at playwright-stealth.new...");
+fs.cpSync(SOURCE, DEST_NEW, { recursive: true });
+
+if (fs.existsSync(DEST_OLD)) {
+	fs.rmSync(DEST_OLD, RM_RETRY_OPTS);
 }
 
-// 4. Simple copy - no symlink handling needed since hoisted mode creates real files
-console.log("Copying to bundle location...");
-fs.cpSync(SOURCE, DEST, { recursive: true });
+console.log("Swapping into place...");
+if (fs.existsSync(DEST)) {
+	fs.renameSync(DEST, DEST_OLD);
+}
+fs.renameSync(DEST_NEW, DEST);
+
+if (fs.existsSync(DEST_OLD)) {
+	fs.rmSync(DEST_OLD, RM_RETRY_OPTS);
+}
 
 console.log("MCP servers prepared successfully.");

